### PR TITLE
Split out MVCC protos into storage/engine

### PR DESCRIPTION
**Pull request by tamird**
Turns out this is pretty non-trivial. See https://github.com/golang/protobuf/pull/42.

### CODE REVIEW

File: proto/status.proto [hunk 1]
```diff
@@ -27,6 +26,33 @@ option (gogoproto.sizer_all) = true;
 option (gogoproto.marshaler_all) = true;
 option (gogoproto.unmarshaler_all) = true;
 
+// MVCCStats tracks byte and instance counts for:
```
Comment (spencerkimball) [inline proto/status.proto]: This doesn't really belong here. It belongs in mvcc.proto. Is there some reason we're not putting it there? Is it impossible to include mvcc.proto in status.proto?

File: build/protobuf.mk [hunk 2]
```diff
@@ -44,13 +44,21 @@ CPP_PROTOS := $(filter %api.proto %data.proto %internal.proto %config.proto %err
 CPP_HEADERS := $(subst ./,$(ENGINE_ROOT)/,$(CPP_PROTOS:%.proto=%.pb.h)) $(subst $(GOGOPROTO_ROOT),$(ENGINE_ROOT),$(GOGOPROTO_PROTO:%.proto=%.pb.h))
 CPP_SOURCES := $(subst ./,$(ENGINE_ROOT)/,$(CPP_PROTOS:%.proto=%.pb.cc)) $(subst $(GOGOPROTO_ROOT),$(ENGINE_ROOT),$(GOGOPROTO_PROTO:%.proto=%.pb.cc))
 
+ENGINE_PROTOS := $(sort $(shell find $(ENGINE_ROOT) -name *.proto -type f))
```
Comment (bdarnell) [inline build/protobuf.mk]: Rename this to ENGINE_CPP_PROTOS unless it's used somewhere in a non-c++ context (which doesn't seem to be the case)

**Comment by spencerkimball** [comment]
LGTM

File: build/protobuf.mk [hunk 3]
```diff
@@ -44,13 +44,21 @@ CPP_PROTOS := $(filter %api.proto %data.proto %internal.proto %config.proto %err
...
 .PHONY: protos
-protos: $(GO_SOURCES) $(CPP_HEADERS) $(CPP_SOURCES)
+protos: $(GO_SOURCES) $(CPP_HEADERS) $(CPP_SOURCES) $(ENGINE_CPP_HEADERS) $(ENGINE_CPP_SOURCES)
 
 $(GO_SOURCES): $(PROTOC) $(GO_PROTOS) $(GOGOPROTO_PROTO) $(PROTOC_PLUGIN)
 	find $(REPO_ROOT) -not -path '*/.*' -name *.pb.go | xargs rm
 	for dir in $(sort $(dir $(GO_PROTOS))); do \
-	  $(PROTOC) -I.:$(GOGOPROTO_PATH) --plugin=$(PROTOC_PLUGIN) --$(PLUGIN_SUFFIX)_out=$(ORG_ROOT) $$dir/*.proto; \
+	  $(PROTOC) -I.:$(GOGOPROTO_PATH) --plugin=$(PROTOC_PLUGIN) --$(PLUGIN_SUFFIX)_out=import_prefix=github.com/cockroachdb/:$(ORG_ROOT) $$dir/*.proto; \
+	  sed -i.bak 's/import math "github\.com\/cockroachdb\/math"//g' $$dir/*.pb.go; \
```
Comment (bdarnell) [inline build/protobuf.mk]: It doesn't look like they add `import_prefix` to the `"math"` import (https://github.com/golang/protobuf/blob/master/protoc-gen-go/generator/generator.go#L1153)

File: proto/status.proto (hunk 1 above)
Comment (tamird) [inline proto/status.proto]: circular dependency.

File: build/protobuf.mk (hunk 3 above)
Comment (tamird) [inline build/protobuf.mk]: we use gogoproto, not golang/protobuf, and i promise i didn't do this for no reason

Comment (tamird) [inline build/protobuf.mk]: note that this doesn't strip the prefix, this removes the line entirely.

File: build/protobuf.mk (hunk 2 above)
Comment (tamird) [inline build/protobuf.mk]: why is that better? there's nothing about these protos that implies they are c++-specific

Comment (bdarnell) [inline build/protobuf.mk]: The protos aren't c++ specific but the way this variable is used is to generate c++ code. These same protos also appear in GO_PROTOS and the go code is generated from their inclusion in that variable, not ENGINE_PROTOS.

**Comment by bdarnell** [comment]
LGTM

File: build/protobuf.mk (hunk 3 above)
Comment (petermattis) [inline build/protobuf.mk]: The `/` in your sed line doesn't need to be a `/`. For example, you could use `,` instead which would remove some of the backslash quoting:

```
  sed -i.bak 's,import math "github\.com/cockroach/math",,g'
```

File: proto/status.proto [hunk 4]
```diff
@@ -27,6 +26,33 @@ option (gogoproto.sizer_all) = true;
...
+//  - Live key/values (i.e. what a scan at current time will reveal;
+//    note that this includes intent keys and values, but not keys and
+//    values with most recent value deleted)
+//  - Key bytes (includes all keys, even those with most recent value deleted)
+//  - Value bytes (includes all versions)
+//  - Key count (count of all keys, including keys with deleted tombstones)
+//  - Value count (all versions, including deleted tombstones)
+//  - Intents (provisional values written during txns)
+//  - System-local key counts and byte totals
+message MVCCStats {
```
Comment (petermattis) [inline proto/status.proto]: Perhaps this should be renamed to `EngineStats` or `StoreStats`.

Comment (tamird) [inline proto/status.proto]: next break-up-protos PR coming in a moment, and it moves this thing again. Want to comment there?

**Pull request merged**